pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java

```diff
@@ -288,43 +293,57 @@ public Response updateFunction(final String tenant, final String namespace, fina
     }
 
     public Response deregisterFunction(final String tenant, final String namespace, final String functionName,
-            String clientRole) {
+            String sourceName, String sinkName, String clientRole) {
```

Comment:
The logic in FunctionImpl is a little hairy.  Maybe the code can be simpler if we just have distinct FunctionImpl, SourceImpl, and SinkImpl?  So that we don't have to check for all three in just one Impl.  Also if we want to do modifications in the future, having separate Impls will allow us to be more certain modifying one thing won't break something else e.g. modifying source won't break functions.

Reply:
I'll add that as a TODO list to break down even the FunctionImpl into those three categories.
For now the abstraction is that from a user perspective, sinks/sources/functions are distintive with their own 'namespaces' in some sense. Currently they use the same backend which doesn;t really distinguish, but atleast from the outside world there is some semblance of difference

Reply:
ok I guess that is fine for right now. I am concerned here since the code logic is complex and bugs can be easily introduced (especially by developers that didn't originally work on this) that will break all three.